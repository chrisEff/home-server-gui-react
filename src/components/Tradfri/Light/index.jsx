'use strict'

const homeServerApi = require('../../../homeServerApi')

const React = require('react')
const PropTypes = require('prop-types')

const LightTitle = require('./LightTitle')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {

	static propTypes = {
		bulb: PropTypes.object.isRequired,
	}

	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<LightTitle title={bulb.name} onChange={this.updateName} />

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateName = async (name) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/name/${name}`)

		bulb.name = name
		this.setState({bulb})
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/state/${state}`)

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/brightness/${brightness}`)

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/color/${color}`)

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
