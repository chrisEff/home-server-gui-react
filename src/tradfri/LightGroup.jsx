const Light = require('./Light')

class LightGroup extends React.Component {
	constructor(props) {
		super(props)
		this.state = {group: props.group}
	}

	render() {
		return (
			<div className='lightGroup'>
				<h3>{this.state.group.name}</h3>
				{
					this.state.group.devices
						.filter(device => device.type === 'bulb')
						.map(device => <Light key={device.name} bulb={device}/>)
				}
			</div>
		)
	}
}

module.exports = LightGroup
