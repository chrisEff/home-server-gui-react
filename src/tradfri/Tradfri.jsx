'use strict'

const config = require('../../config')
const LightGroup = require('./LightGroup')
const ErrorMessage = require('../ErrorMessage')

class Tradfri extends React.Component {
	constructor (props) {
		super(props)
		this.state = {
			groups: [],
			errorMsg: null,
		}
	}

	async componentDidMount () {
		try {
			const groups = await (await fetch(`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/group?key=${config.api.key}`)).json()
			const devices = await (await fetch(`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device?key=${config.api.key}`)).json()

			groups.forEach(group => {
				group.devices = []
				group.deviceIds.forEach(deviceId => {
					const device = devices.find(device => device.id === deviceId)
					group.devices.push(device)
				})
				group.devices.sort((a, b) => {
					return a.name > b.name ? 1 : b.name < a.name ? -1 : 0;
				})
			})
			groups.sort((a, b) => {
				return a.name > b.name ? 1 : b.name < a.name ? -1 : 0;
			})

			this.setState({
				groups
			})
		} catch (e) {
			this.setState({errorMsg: e.message})
		}
	}

	render () {
		return (
			<div>
				<h2>Licht</h2>
				{this.state.errorMsg && <ErrorMessage message={this.state.errorMsg}/>}
				{this.state.groups.map(group => <LightGroup key={group.name} group={group}/>)}
			</div>
		)
	}
}

module.exports = Tradfri
