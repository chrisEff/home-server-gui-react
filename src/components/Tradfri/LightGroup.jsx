'use strict'

import React, {useRef, useState} from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {StyleSheet, css} from 'aphrodite'

import Light from './Light'
import {setGroupName} from '@/actions/tradfri'

const LightGroup = ({id, name, devices, onSaveName}) => {
	const [editMode, setEditMode] = useState(false)
	const nameInput = useRef(name)

	return (
		<div className={css([styles.lightGroup])}>
			{!editMode && <h3  className={css([styles.name])} onClick={() => { setEditMode(true) }}>{name}</h3>}
			{editMode && <div>
				<input ref={nameInput} defaultValue={name} />
				<button onClick={() => { setEditMode(false); onSaveName(id, nameInput.current.value) }}>OK</button>
			</div>}

			{
				devices
					.filter(device => device.type === 'bulb')
					.sort((a, b) => a.name.localeCompare(b.name))
					.map(device => <Light key={device.name} id={device.id} />)
			}
		</div>
	)
}

LightGroup.propTypes = {
	id: PropTypes.number.isRequired,
	name: PropTypes.string.isRequired,
	devices: PropTypes.array.isRequired,
	onSaveName: PropTypes.func,
}

const styles = StyleSheet.create({
	lightGroup: {
		display: 'inline-block',
	},
	name: {
		':hover': {
			':after': {
				content: "' ✏️'",
				fontSize: '0.9em',
			},
		},
	},
})

const mapStateToProps = (state, ownProps) => ({
	devices: state.tradfriDevices.filter(device => ownProps.deviceIds.includes(device.id)),
})

const mapDispatchToProps = dispatch => ({
	onSaveName: (id, name) => dispatch(setGroupName(id, name)),
})

export default connect(mapStateToProps, mapDispatchToProps)(LightGroup)
export {LightGroup}
