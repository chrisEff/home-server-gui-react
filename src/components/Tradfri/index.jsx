'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import LightGroup from './LightGroup'
import {loadTradfriDevices, loadTradfriGroups} from '../../actions'

class Tradfri extends React.Component {

	static propTypes = {
		title: PropTypes.string,
	}

	async componentDidMount () {
		this.props.onLoad()
	}

	render () {
		return (
			<div id='tradfri'>
				<h2>{this.props.title}</h2>
				{this.props.tradfriGroups.map(group => <LightGroup key={group.name} id={group.id} name={group.name} deviceIds={group.deviceIds}/>)}
			</div>
		)
	}
}

const mapStateToProps = state => ({
	tradfriGroups: state.tradfriGroups,
	tradfriDevices: state.tradfriDevices,
})

const mapDispatchToProps = dispatch => ({
	onLoad: () => {
		dispatch(loadTradfriGroups())
		dispatch(loadTradfriDevices())
	},
})

export default connect(mapStateToProps, mapDispatchToProps)(Tradfri)
export {Tradfri}
