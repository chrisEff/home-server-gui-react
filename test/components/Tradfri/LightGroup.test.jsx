'use strict'

import React from 'react'
import {LightGroup} from 'src/components/Tradfri/LightGroup'

import {shallow} from 'enzyme'

describe('LightGroup', () => {
	it('should render correctly', () => {
		const wrapper = shallow(<LightGroup id={1} name='Test Group' devices={[{
			brightness: 254,
			bulbType: 'white-spectrum',
			color: 'neutraÃ¥l',
			firmware: '1.2.217',
			id: 65537,
			model: 'TRADFRI bulb E27 WS opal 980lm',
			name: 'Decke',
			state: 0,
			type: 'bulb',
		}]} />)
		expect(wrapper).toMatchSnapshot()
	})
})
