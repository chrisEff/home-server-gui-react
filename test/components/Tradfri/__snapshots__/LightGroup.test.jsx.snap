// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LightGroup should go into edit mode on click 1`] = `
<div
  className="lightGroup_36rlri"
>
  <div>
    <input
      defaultValue="Test Group"
    />
    <button
      onClick={[Function]}
    >
      OK
    </button>
  </div>
  <Connect(Light)
    id={65537}
    key="Decke"
  />
</div>
`;

exports[`LightGroup should render correctly 1`] = `
<div
  className="lightGroup_36rlri"
>
  <h3
    className="name_7bdam5"
    onClick={[Function]}
  >
    Test Group
  </h3>
  <Connect(Light)
    id={65537}
    key="Decke"
  />
</div>
`;
